design-proposals/host-ip-pool.md

```diff
@@ -0,0 +1,349 @@
+
+**cluster admin** is a human user responsible for managing the cluster node
+networking.
+
+1. The cluster admin create `NodeNetworkIPPool` resources with IP ranges and a names.
+2. The cluster admin configure node network interfaces statically but with IP from the `NodeNetworkIPPool` using a
+`NodeNetworkConfigurationPolicy` referencing them.
+3. The cluster admin wait for the `NodeNetworkConfigurationPolicy` to success
+
+#### Variation (not free IP address at the NodeNetworkIPPool)
+
+If there is not a free IP address at referenced pool NNCP will fail with a explicit message, so the range has to be
+increased and NNCP applied again
+
+### Workflow Description (removing a reference to NodeNetworkIPPool)
+
+**cluster admin** is a human user responsible for managing the cluster node
+networking.
+
+1. The cluster admin changes network interface to remove the refernece to the NodeNetworkIPPool (for example harcoding the IP or switching to DHCP).
+2. The cluster admin wait for the NodeNetworkConfigurationPolicy to success so the IPs from that interfaces are free now.
+
+### Workflow Description (scaling up cluster)
+
+**cluster admin** is a human user responsible for managing the cluster node
+networking.
+
+1. The cluster admin add a new node to the cluster already configured using IPools
+2. The cluster admin wait for the NodeNetworkConfigurationPolicy to success so the new node has took address from the NodeNetworkIPPool
+
+
+### Workflow Description (scaling down cluster)
+
+**cluster admin** is a human user responsible for managing the cluster node
+networking.
+
+1. The cluster admin remove a node form a cluster already configured using `NodeNetworkIPool`
+2. The cluster admin wait for the `NodeNetworkConfigurationPolicy` to success so the IPs from removed node are free.
+
+### Workflow Description (manual NodeNetworkIPPool clean up)
+
+**cluster admin** is a human user responsible for managing the cluster node
+networking.
+
+1. The cluster admin find that some IPs are not used anymore but still allocated
+2. The cluster admin reset `NodeNetworkIPPool` allocated status
+4. The cluster admin wait for `NodeNetworkIPPool` status to success
+
+NOTE: This is a manual garbage collector that is good enough for first implementation it can be converted into automatic in the future.
+
+### API Extensions day 2
+
+This proposal add a new CRD to configure and store IP pools (NodeNetworkIPPool) and some
+changes at the NNCPs to reference the ippool.
+
+Following is an example CR for the cluster wide CRD `NodeNetworkIPPool`:
+
+```yaml
+apiVersion: nmstate.io/v1
+kind: NodeNetworkIPPool
+metadata:
+  name: traffic-1
+spec:
+  range:
+    cidr: 10.10.10.0/24
+    start: 1
+    end: 10
+    exclude: [4, 7]
+status:
+  allocated-offsets: [1, 2, 5, 6, 8]
+  conditions:
+  - type: Available
+    status: True
+    reason: ReconciledPool
+    message: The pool is up to date
+  - type: Failure
+    status: False
+    reason: ReconciledPool
```

Comment:
This are not events this are conditions, and they denote a final state, a user can wait for a Failure condition and different user can wait for a Available condition:
```bash
kubectl wait --for=condition=Available
kubectl wait --for=condition=Failure
```

In this case we are saying that Failure is false since the pool is reconciled.

Reply:
Right, keeping just the Available condition with proper Reason and adding a pair of example with the two errors.